cloudpickle/cloudpickle.py

```diff
@@ -694,7 +705,11 @@ def _function_getstate(func):
     #   unpickling time by iterating over slotstate and calling setattr(func,
     #   slotname, slotvalue)
     slotstate = {
-        "__name__": func.__name__,
+        # Create a copy of the function name, to avoid memoization. This is necessary
+        # to ensure deterministic pickles, that depends wheter the name is interned
+        # or not. This is not guaranteed for reconstructed dynamic function, so we make
+        # sure it is never interned.
+        "__name__": "".join(func.__name__),
         "__qualname__": func.__qualname__,
```

Comment:
Any idea why we don't need this treatment for the `__qualname__` value nor for the keys of `func.__dict__`?

Reply:
Actually, at least for the latter, there are still things to solver. I will push a slight change to the test to show what I mean.

Reply:
The issue with string interning is mostly due to collision between the string defined in two different places: typically method names that are interned as class attributes, vs the dynamic func name.
So I would say this won't happen for `__qualname__` as this is only present here but maybe I am wrong.